scripts: add a change log generator script

knz opened the pull request:
This groups commits by author, in the style of the Linux CHANGES file.

For example:

```
% ./scripts/changelog.py v1.0.5 v1.0.4
Ben Darnell:
        #17654 182613 (+  63 -  20 ~  83/ 1) cherrypick: security: Disable 3DES cipher suites

Jordan Lewis:
        #17653 828a46 (+  23 -   4 ~  27/ 4) cherrypick: sql: two bugfixes to fks that use a partial index (2 commits)

kena:
        #17689 6255bb (+   7 -   2 ~   9/ 1) cherrypick: log: allow panic to propagate on crash reporting timeouts
        #17675 a725c2 (+  17 -   5 ~  22/ 2) cherrypick: log: ensure Fatal calls are reported even if logs don't go to stderr
        #17681 e73fc3 (+  50 -  39 ~  89/ 2) cherrypick: sql: disallow AS OF SYSTEM TIME 0, add more AS OF tests
        #17281 8d0dea (+  15 -   0 ~  15/ 3) cherrypick: sql: ensure that planNodes are Close()d on error after expansion

Tobias Schottdorf:
        #17385 63abc8 (+  18 -   3 ~  21/ 1) cherry-pick: base: bump DefaultSendNextTimeout from 500ms to 10m
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17988)
<!-- Reviewable:end -->

knz committed:
scripts: add a change log generator script

This groups commits by author, in the style of the Linux CHANGES file.

For example:

```
% ./scripts/changelog.py v1.0.5 v1.0.4
Ben Darnell:
        #17654 182613 (+  63 -  20 ~  83/ 1) cherrypick: security: Disable 3DES cipher suites

Jordan Lewis:
        #17653 828a46 (+  23 -   4 ~  27/ 4) cherrypick: sql: two bugfixes to fks that use a partial index (2 commits)

kena:
        #17689 6255bb (+   7 -   2 ~   9/ 1) cherrypick: log: allow panic to propagate on crash reporting timeouts
        #17675 a725c2 (+  17 -   5 ~  22/ 2) cherrypick: log: ensure Fatal calls are reported even if logs don't go to stderr
        #17681 e73fc3 (+  50 -  39 ~  89/ 2) cherrypick: sql: disallow AS OF SYSTEM TIME 0, add more AS OF tests
        #17281 8d0dea (+  15 -   0 ~  15/ 3) cherrypick: sql: ensure that planNodes are Close()d on error after expansion

Tobias Schottdorf:
        #17385 63abc8 (+  18 -   3 ~  21/ 1) cherry-pick: base: bump DefaultSendNextTimeout from 500ms to 10m
```

benesch commented:
Are you planning to check in the resultant changelog, @knz?

dt commented:
last time we discussed this, I recall finding Angular's commit message format + change log generator a potentially interesting example: https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#

knz commented:
@benesch I'm not sure why we would need to commit files that can be auto-generated. Otherwise, I will indeed suggest to append these change logs to the release announcements.

@dt the Angular change log format is *much* too verbose to my test. Also :bike: :house_with_garden:

bdarnell commented:
Are you suggesting this in addition to the current release notes or instead of them? This works OK for patch releases, but is very overwhelming in larger ones. 

I don't think grouping by author makes it more intelligible - seeing e.g. changes related to backup/restore spread across the several members of that team would just be confusing.

If we're going to publish this to users it would be primarily in a web-based form, so we'd want links, etc. Github [can do this](https://github.com/cockroachdb/cockroach/compare/v1.0.4...v1.0.5) in a way that's not perfect but might be close enough.

knz commented:
So I want to merge this PR.
We already have various scripts useful to different people in `scripts/` and never before the discussion has come up "What should the UX be" and "are we going to publish the output".

There are two questions that are open here:
1) is there a space in `scripts/` for this one yes/no (if the answer is "no" then please suggest in which repo I can park it instead; and then I will also open the other scripts to the question)
2) is the code reasonable or do you foresee unexpected/surprising breakage (same quality standard for a review as you would apply to the other scripts in that directory)

Thank you.

benesch commented:
From past experience, the `scripts/` directory can devolve into a dumping ground of infrequently-used, poorly-maintained scripts, each used by just a handful of people. I think `scripts/authors.sh` and `scripts/git-blame-stats.sh` are guilty of this already. There's not much Cockroach-specific about them. I share @bdarnell's concern that this script won't be generally useful.

All that said, I get the discoverability argument. Way harder to find these things if they live elsewhere. Perhaps the `scripts` directory just needs some more order imposed.

Anyway, I'll defer to @bdarnell. The code itself LGTM, except that I wonder if you could repurpose the existing `.mailmap` instead of hardcoding author aliases?

bdarnell commented:
-1 on checking this in to the main repo as-is (and i'd be +1 on removing several of the scripts that are in this directory now). The closest thing we currently have to a place for this is [on the wiki](https://github.com/cockroachdb/cockroach/wiki/Rando-scripts). 

If we want to move towards auto-generating release notes from commit logs, that's another story (I'm interested in getting out of the release notes business). But in that case I think we want a format that's rather different from the linux-style one here (and an agreement across the team to make sure that our commit messages are more comprehensible for end users). 

If you want to go down that path, my feedback on the format is
- emphasize the description of the change. I'd drop the diffstats and author completely.
- group by package/topic, not author
- emit markdown or html (or rst when/if docs switches), with links
- try to strip boilerplate like the word "cherrypick"

We don't have to keep following the exact format I've been using, but I think we want something closer to it than to the mechanics-focused linux format.

knz commented:
Closing in favor of #19309 / #19308.
